File: LeetCode/Python Solutions/Even Odd Tree/even.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isEvenOddTree(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return False
        
        from collections import deque
        queue = deque([(root, 0)])  # Node paired with its level
        
        while queue:
            prev_value = None
            for _ in range(len(queue)):
                node, level = queue.popleft()
                if level % 2 == 0:  # Even level: values must be odd and strictly increasing
                    if node.val % 2 == 0 or (prev_value is not None and node.val <= prev_value):
                        return False
                else:  # Odd level: values must be even and strictly decreasing
                    if node.val % 2 != 0 or (prev_value is not None and node.val >= prev_value):
                        return False
                prev_value = node.val
                
                if node.left:
                    queue.append((node.left, level + 1))
                if node.right:
                    queue.append((node.right, level + 1))
        
        return True

File: LeetCode/Java Solutions/Even Odd Tree/even.java
public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public boolean isEvenOddTree(TreeNode root) {
        if (root == null) {
            return false;
        }
        
        Queue<Pair<TreeNode, Integer>> queue = new LinkedList<>();
        queue.add(new Pair<>(root, 0));  // Node paired with its level
        
        while (!queue.isEmpty()) {
            Integer prevValue = null;
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                Pair<TreeNode, Integer> pair = queue.poll();
                TreeNode node = pair.getKey();
                int level = pair.getValue();
                
                if (level % 2 == 0) {  // Even level: values must be odd and strictly increasing
                    if (node.val % 2 == 0 || (prevValue != null && node.val <= prevValue)) {
                        return false;
                    }
                } else {  // Odd level: values must be even and strictly decreasing
                    if (node.val % 2 != 0 || (prevValue != null && node.val >= prevValue)) {
                        return false;
                    }
                }
                prevValue = node.val;
                
                if (node.left != null) {
                    queue.add(new Pair<>(node.left, level + 1));
                }
                if (node.right != null) {
                    queue.add(new Pair<>(node.right, level + 1));
                }
            }
        }
        
        return true;
    }
}
